### ISSUE TITLE
Downcast a Gtk.Widget to a concrete class

### ISSUE THREAD

**Issue by haijin-development** [romgrk/node-gtk#126]
Hi.
I want to add a context menu to a Gtk.TextView using the following event hook:

```javascript
textView.on('populate-popup', (menu) => {
    menu.add(new Gtk.MenuItem())
})
```
The problem is that `menu` comes as an abstract `Gtk.Widget` and does not recognize the `.add()` function.

Where should I modify the library to downcast the menu to a `Gtk.Menu` object?

Thanks!

**Comment by haijin-development** [comment]
Actually it would nice to be able to cast any object pointer to any pointer type from the node library to allow to work with objects that are not even yet wrapped by the library.

```javascript
Gtk.castTo = function (gtkObject, gtkType) {
    return (gtkType) gtkObject;
}
```

And the previous example would look like

```javascript
textView.on('populate-popup', (menuHandle) => {
    const menu = Gtk.castTo(menuHandle, Gtk.Menu)
    menu.add(new Gtk.MenuItem())
})
```

Would something like that be possible?

**Comment by haijin-development** [comment]
One more question before closing this issue.

The following workaround does work:

```javascript
textView.on('populate-popup', (gtkWidget) => {
    const menu = new Gtk.Menu()
    gtkWidget.add = menu.add.bind(gtkWidget)
    gtkWidget.showAll = menu.showAll.bind(gtkWidget)

    const menuItem = new Gtk.MenuItem({label: 'New menu item'})

    gtkWidget.add(menuItem)
    gtkWidget.showAll()
})
```

but what is the proper way to do that in Node?

Thanks in advance.

**Comment by romgrk** [comment]
Hey, thanks for the report.

So indeed, according to the [documentation](https://developer.gnome.org/gtk3/stable/GtkTextView.html#GtkTextView-populate-popup) for the `populate-popup` signal, it's a `GtkWidget`, therefore `node-gtk` will provide it as such.

### How to cast it

```js
gtkWidget.__proto__ = Gtk.Menu.prototype
// then
gtkWidget.add(menuItem)
```
### How to fix the issue

I think `gtkWidget.__gtype__` should contain the GType of the correct type, `GtkMenu`. Therefore, we could do something like we do [here](https://github.com/romgrk/node-gtk/blob/master/lib/overrides/Gtk-3.0.js#L87-L100) in `node-gtk` before calling the callback. We would still be respecting the type, because `GtkMenu` is a `GtkWidget`. Unfortunately, I'm very low on time these days, I'm not sure when I'll be able to implement it. If you'd like to open a PR for this, I can provide guidance as to where the faulty code is.

**Comment by haijin-development** [comment]
Now I have implemented it properly, thanks!

I tried to find a place to hook the event callbacks before they are `call`ed to perform the down cast for the `populate-popup` event but I could not find it in the node side. Are the event callbacks activated on the C side?

Yes, if you could point the place to hook the event callbacks I might try to implement it and create a PR.

Thanks!

**Comment by romgrk** [comment]
Yes, signals (aka events) callbacks are called only on the `C++` side.

[This function](https://github.com/romgrk/node-gtk/blob/master/src/closure.cc#L21-L82) is the one called when there is a signal dispatched. It takes the parameters from GLib, transforms those to JS values, then calls the event handler on line 59.

The issue could also be solved in a different place. As you can see on line 53 of that function, there is a call to `GIArgumentToV8(...)`, which is the function that makes the `GIArgument` to JS value transformation. You could modify that part to return objects with their most specific type (`GtkMenu` in this example) instead of the type passed through `&type_info`.

Now that I see that function itself defers to `WrapperFromGObject` [here](https://github.com/romgrk/node-gtk/blob/master/src/value.cc#L123), which is in charge of transforming a GObject to a JS Object. `WrapperFromGObject` is defined [here](https://github.com/romgrk/node-gtk/blob/master/src/gobject.cc#L414). To fix the issue, you would need to replace the `GIBaseInfo object_info` [here](https://github.com/romgrk/node-gtk/blob/master/src/gobject.cc#L431) by getting the `GIBaseInfo` of the GObject with [`g_irepository_find_by_gtype`](https://developer.gnome.org/gi/stable/GIRepository.html#g-irepository-find-by-gtype). The GType of a GObject can be obtained with [`G_OBJECT_TYPE`](https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html#G-OBJECT-TYPE:CAPS)

I think you have enough information with that, let me know if you have more questions :)

**Comment by WebReflection** [comment]
FWIW `gtkWidget.add(Object.setPrototypeOf(menuItem, Gtk.Menu.prototype))` is a better way to workaround this, since `__proto__` is both annex B of the specs and unreliable for objects not inheriting `Object.prototupe`, hence the best way to go.

**Comment by haijin-development** [comment]
Thank you both. I will try that soon.

As a funny note, I just realized how easy it is to inject code from a malicious npm package into pretty much any level of a node app just by modifying the prototype of a function parameter like, say, a String.
